### ISSUE TITLE
phoityne-vscode very slow to initialise

### ISSUE THREAD

**Issue by mpickering** [phoityne/phoityne-vscode#45]
I am trying to load GHC into phoityne.

Loading GHC into GHCI doesn't take very long, < 1 minute.

Using the same command with `phoityne` leads to a period of about 5 minutes where `phoityne` uses a lot of CPU resources and no output is displayed to the terminal. After this the output of running the command to load GHC into GHCi is displayed. 

What is `phoityne` doing in those initial 5 minutes before it starts to load ghci?

**Comment by phoityne** [comment]
Hi.

> What is phoityne doing in those initial 5 minutes before it starts to load ghci?

Doing almost nothing before starting ghci.
Although there are two times communication with vscode.

Just run the command defined in the launch.json ("ghciCmd" attribute).
After launched ghci, :load the start file defined as "startup" attribute in the launch.json.
This will take some time because of loading all related modules.

Regards.

**Comment by mpickering** [comment]
Perhaps this is because the command is run using `bash ./hadrian/ghci.sh` rather than running the command directly? It's not too bad so I will debug this myself if 

Another issue. The debugger hangs very soon after starting, I think it's because `ghc` tries to read something from stdin? What is the workaround for this, or at least how to find the call which causes the process to hang?

**Comment by phoityne** [comment]
Could you let me know below information ?

- vscode version
- ghc version
- which os?
- using stack?
- launch.json
- console log with level "DEBUG" (setting logLevel in launch.json)

**Comment by mpickering** [comment]
I have an exerimental setup so the details aren't going to be much use to reproduce this issue. 

I am using a custom build of GHC HEAD and then setting the ghci script to the custom one (`./hadrian/ghci.sh`) which loads ghc into ghci. The debugger works correctly when used directly inside of ghci but invoking `:main` from within the debug console causes the hang. I strongly suspect this is due to trying to read from `stdin` or something like that but I don't know how to verify or debug this.

**Comment by phoityne** [comment]
I'm afraid that I don't understand what you are trying to do.
phoityne-vscode.exe is not a ghci nor a debugger.

Are you trying to run "phoityne-vscode main" from ghci prompt ?
phoityne-vscode will wait json request from stdin, so ghci will hang.

**Comment by mpickering** [comment]
The GHCi debugger works for any project you can load into GHCi. There is a script which loads GHC itself into GHCi so developers can use the repl features. This is invoked using `./hadrian/ghci.sh`. 

My understanding was that `phoityne-vscode` worked with any project which you could load into GHCi. I tested my setup with `phoityne-vscode` on a simple project which worked fine. Then I tried to modify the `launch.json` so that the `ghciCmd` was set to `./hadrian/ghci.sh`. This loaded the repl correctly but then hung after running `:trace main`. 

So I suspect that `:trace main` hangs because of something to do with stdin/stderr but I don't know how to debug which part of GHC would affect this, how to work around this or what to patch in GHC to stop it accessing stdin.

**Comment by phoityne** [comment]
> Then I tried to modify the launch.json so that the ghciCmd was set to ./hadrian/ghci.sh. This loaded the repl correctly but then hung after running :trace main.

So, this ":trace main" means tracing ghc main below ?
  https://github.com/ghc/ghc/blob/master/ghc/Main.hs
```
  L91 main :: IO ()
  L92 main = do
  L93    initGCStatistics -- See Note [-Bsymbolic and hooks]
```

Could you put breakpoint on L93 and can break there ?

**Comment by mpickering** [comment]
Yes exactly. I can put a breakpoint there and it breaks but then continuing causes hangs. Perhaps the `hSetBuffering` causes issues as well?

**Comment by phoityne** [comment]
>  Perhaps the hSetBuffering causes issues as well?

Isn't it able to step(by F11) after breaking at L93 ?
If it hangs immediately, hSetBuffering might be  the cause.
